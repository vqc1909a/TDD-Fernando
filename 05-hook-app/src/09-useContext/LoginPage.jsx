import { useContext } from "react"
import { UserContext } from "./context/UserContext";

export const LoginPage = () => {

  const {user, setUser} = useContext(UserContext);
  return (
    <>
        <h1>LoginPage</h1>
        <hr />
        <pre aria-label="pre_login">
          { JSON.stringify(user, null, 3)}
        </pre>

        <button aria-label="button_login" className="btn btn-primary" onClick={() => setUser({id: 123, name: 'Juan', email: 'juan@google.com'})}>Set user</button>
    </>   
  )
}
