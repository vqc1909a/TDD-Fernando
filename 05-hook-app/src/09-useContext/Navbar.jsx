import { NavLink } from "react-router-dom"

export const Navbar = () => {
  return (
    <nav className="navbar navbar-expand-lg navbar-dark bg-dark">
        <div className="container-fluid">
            <NavLink className="navbar-brand" to="/">USE CONTEXT</NavLink>

            <div className="collapse navbar-collapse" id="navbarNav">
                <ul className="navbar-nav">
                    <li className="nav-item">
                        <NavLink className={(args) => {
                            // console.log(args)
                            // { isActive: true, isPending: false, isTransitioning: false }
                            return `nav-link ${args.isActive ? 'active' : ''}`
                        }} to="/">Home</NavLink>
                    </li>

                    <li className="nav-item">
                        <NavLink className={(args) => {
                            // console.log(args)
                            return `nav-link ${args.isActive ? 'active' : ''}`
                        }} to="/about">About</NavLink>
                    </li>
                    <li className="nav-item">
                        {/* We use the componente NavLink instead use Link because we can add the class "active" when the link has matched with the path url  */}
                        <NavLink className={(args) => {
                            // console.log(args)
                            return `nav-link ${args.isActive ? 'active' : ''}`
                        }
                        } to="/login">Login</NavLink>
                    </li>
                </ul>
            </div>
        </div>
    </nav>
)
}
