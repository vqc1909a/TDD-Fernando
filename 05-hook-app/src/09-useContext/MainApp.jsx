import { Navigate, Route, Routes } from "react-router-dom";
import { AboutPage, HomePage, LoginPage } from "./";
import { Navbar } from "./Navbar";
import { UserProvider } from "./context/UserProvider";



export const MainApp = () => {
  return (
    // This UserProvider will should be in the rootComponent called main.jsx, it is here only for the example
    <UserProvider>

        <h1>MainApp</h1>
        <Navbar />
        <hr />

        <Routes>
          <Route path="/" element={<HomePage />} />
          <Route path="about" element={<AboutPage />} />
          <Route path="login" element={<LoginPage />} />
          <Route path="/*" element={<Navigate to="/about" />}/>
        </Routes>
    </UserProvider>
  )
}
