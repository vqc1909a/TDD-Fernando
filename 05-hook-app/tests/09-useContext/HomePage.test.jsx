import { render, screen } from "@testing-library/react"
import { HomePage } from "../../src/09-useContext/HomePage"
import { UserContext } from "../../src/09-useContext/context/UserContext";

describe('tests on HomePage component', () => { 

  const user = {
    id: 1,
    name: 'vqc1909a',
    email: 'vqc190a@gmail.com'
  }

  test('should show the component without user', () => { 
    //In one component that used useContext, we need to envolve the test component with the component provider of the context
    const { container } = render(
      <UserContext.Provider value={{user: undefined}}>
        <HomePage />
      </UserContext.Provider>
    );
    const preTag = screen.getByLabelText('pre');
    expect(preTag.textContent).toBeFalsy();
    expect(preTag.textContent).toBe("");
    expect(container).toMatchSnapshot();
  })

  test('should show the component with user', () => { 
    //In one component that used useContext, we need to envolve the component to test with the component provider of the context
    render(
      <UserContext.Provider value={{user}}>
        <HomePage />
      </UserContext.Provider>
    );
    const preTag = screen.getByLabelText('pre');
    expect(preTag.textContent).toContain(user.name);
    expect(preTag.textContent).toContain(user.id.toString());

  })
})