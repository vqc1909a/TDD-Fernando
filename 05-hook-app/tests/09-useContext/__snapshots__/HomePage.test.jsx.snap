// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests on HomePage component should show the component without user 1`] = `
<div>
  <h1>
    HomePage 
  </h1>
  <hr />
  <pre
    aria-label="pre"
  />
</div>
`;
